src/RA_Dlg_MemBookmark.h

```diff
@@ -8,93 +8,169 @@
+    _NODISCARD static INT_PTR CALLBACK s_MemBookmarkDialogProc(_In_ HWND, _In_ UINT,
+        _In_ WPARAM, _In_ LPARAM);
+    _NODISCARD INT_PTR CALLBACK MemBookmarkDialogProc(_In_ HWND, _In_ UINT,
+        _In_ WPARAM, _In_ LPARAM);
 
-    void InstallHWND(HWND hWnd) { m_hMemBookmarkDialog = hWnd; }
-    HWND GetHWND() const { return m_hMemBookmarkDialog; }
-    BOOL IsActive() const;
+    void InstallHWND(_In_ HWND hWnd) { m_hMemBookmarkDialog = hWnd; }
+    _NODISCARD _CONSTANT_FN& GetHWND() const { return m_hMemBookmarkDialog; }
+    _NODISCARD BOOL IsActive() const;
 
-    std::vector<MemBookmark*> Bookmarks() { return m_vBookmarks; }
-    void UpdateBookmarks(bool bForceWrite);
-    void AddBookmark(MemBookmark* newBookmark) { m_vBookmarks.push_back(newBookmark); }
-    void WriteFrozenValue(const MemBookmark& Bookmark);
+    // Needs non-const reference version for EditProcBM
+    _NODISCARD auto& Bookmarks() { return m_vBookmarks; }
+    _NODISCARD auto& Bookmarks() const { return m_vBookmarks; }
+
+    _NODISCARD auto end() const noexcept
+        ->decltype(std::declval<MyBookmarkMap&>().cend())
+    {
+        return m_BookmarkMap.cend();
+    }
+
+    void UpdateBookmarks(_In_ bool bForceWrite);
+    auto AddBookmark(MemBookmark&& newBookmark) noexcept // ok it's void
+        ->decltype(std::declval<MyBookmarks&>().push_back(std::move(newBookmark)))
+    {
+        MemBookmark myClone;
+        newBookmark.Clone(myClone);
+        m_vBookmarks.push_back(std::move(newBookmark));
+        AddBookmarkMap(std::move(myClone));
+
+    }
+    void WriteFrozenValue(_In_ const MemBookmark& Bookmark);
 
     void OnLoad_NewRom();
 
 private:
-    int m_nNumOccupiedRows;
+    int m_nNumOccupiedRows{ 0 };
 
     void PopulateList();
-    void SetupColumns(HWND hList);
-    BOOL EditLabel(int nItem, int nSubItem);
+    void SetupColumns(_In_ HWND hList);
+    _NODISCARD BOOL EditLabel(_In_ int nItem, _In_ int nSubItem);
     void AddAddress();
-    void ClearAllBookmarks();
-    unsigned int GetMemory(unsigned int nAddr, int type);
+    void ClearAllBookmarks() noexcept;
+    _NODISCARD unsigned int GetMemory(_In_ unsigned int nAddr, _In_ int type);
 
     void ExportJSON();
-    void ImportFromFile(std::string filename);
-    void GenerateResizes(HWND hDlg);
-    std::string ImportDialog();
+    void ImportFromFile(_In_ std::string&& filename);
+    void GenerateResizes(_Inout_ HWND& hDlg);
+    _NODISCARD std::string ImportDialog();
 
-    void AddBookmarkMap(MemBookmark* bookmark)
-    {
-        if (m_BookmarkMap.find(bookmark->Address()) == m_BookmarkMap.end())
-            m_BookmarkMap.insert(std::map<ByteAddress, std::vector<const MemBookmark*>>::value_type(bookmark->Address(), std::vector<const MemBookmark*>()));
-
-        std::vector<const MemBookmark*> *v = &m_BookmarkMap[bookmark->Address()];
-        v->push_back(bookmark);
-    }
+    void AddBookmarkMap(_In_ MemBookmark&& bookmark) noexcept;
```

Comment:
Put some annotations and ran SAL. All of these failed [[noreturn]], just using return; makes the app terminate. Might also be a bug because sometimes doesn't throw as well.